### ISSUE TITLE
LLVM on Aarch64 Tracker

### ISSUE THREAD

**Issue by kd-11** [RPCS3/rpcs3#15567]
LLVM doesn't work properly on aarch64. There are architectural reasons for this, particularly when used within RPCS3. There is some internal progress for all of these, so we're tracking things here to collaborate more efficiently.

### Pre
- [x] Re-implement VM escape on aarch46 using far-jump and context save/restore.

### Main
- [x] PPU LLVM
    - [x] Reimplement call gate using register context only.
    - [x] Diagnose mysterious ret path causing safety asserts to be hit
    - [x] Fix commercial games crashing with nullptr access.
    - [x] Refactoring/cleanup (guest/host context switches, to be shared with SPU)
    - [x] Identify and fix regression after rebase on recent master
 
    <details><summary><h3>Obsolete implementation</h3></summary>

    - [x] Breaking from guest to hypervisor (escape) is broken because of callstack unwinding being incompatible with LLVM's GHC implementation on aarch64.
        - [x] Worked around by having a manual call stack.
    - [x] LLVM is clobbering the link register in GHC blocks making them noreturn.
        - [x] Worked around by modifying LLVM's reserved register list for GHC to at least leave the LR alone.
        - [x] File a report with upstream with generated blocks.
     </details>
- [x] SPU LLVM
    - [x] Same escape issue as PPU LLVM
    - [x] Compilation failures
    - [x] Stack smashing
    - [x] Stack underrun
    - [x] Reimplement unwind + stack scratch hacks as function transformation passes and remove inline asm hooks.
    - [x] Properly formalize/refactor guest-host(hypervisor) context switches and have the code in one place. Currently littered everywhere.
    - [x] Fix lockup after rebase on current master
    - [x] ~~Basic optimizations (call, shufb)~~ (deferred)
    - [x] ~~Other undiagnosed issues (list will be added)~~

**Comment by kd-11** [comment]
I had an epiphany about the LLVM incompatibilities. I'm fighting the ABI so hard trying to make everything stack-based on an architecture that is all about registers. There is 0 reason for me to do that. The stack doesn't need any unwinding or preserving at all.
I can unwind by just doing a far jump to the dispatcher to simulate the ret chain. It's also much faster and doesn't require a hacked up LLVM branch to work.

**Comment by kd-11** [comment]
Well, we're back to running simple games after the guest enter/exit rewrite.
![image](https://github.com/RPCS3/rpcs3/assets/15904127/c226e46a-f03f-48fb-b260-896a1b2dd6cd)

Now comes the boring task of figuring out why most commercial games crash.
The good news is that the new approach does not require a hacked LLVM anymore and upstream works fine. However, I still need to solve the random jumps to uninitialized memory before we can see a PR.

**Comment by kd-11** [comment]
Commercial games now work with PPU LLVM though there is some severe memory corruption coming from the LLVM-generated subroutines. The odd GHC behavior continues and I can see functions writing to stack-based addresses without first allocating memory. A workaround is to just allow a working area on the stack on the gateway. The execution context is no longer saved on the stack to avoid random corruption as well.
Next is to get SPU up to speed, refactor and port the changes on top of modern rpcs3.

**Comment by kd-11** [comment]
And now we're back to needing a custom LLVM again. My original work was based on a specific branch of rpcs3 that was very old (maybe 1-2 years) since that was the first time I started playing with it. The current emulator code doesn't work so well.
I have a workaround in mind, but it may be a long time before I have a chance to touch this subproject again.

**Comment by kd-11** [comment]
SPU + PPU running native LLVM.
![image](https://github.com/RPCS3/rpcs3/assets/15904127/3f34871c-cdb7-4d08-b605-0f2797105ad7)

The codegen is unoptimized, there is a lot of room for improvement. I'm never writing a MachineFunctionPass again so I hope I can just optimize in IR.

**Comment by kd-11** [comment]
With research being completed, I've created some smaller tasks to help merge in the changes. I'll close this ticket once the other 4 have been closed.

**Comment by kd-11** [comment]
Closing now that we have shipping CI builds. It's now official 🥳
